### PR TITLE
Homework week6-try 6

### PR DESCRIPTION
**Pull request by anilsukul**
Processed feedback from Maciek (added more testcases, no allocation in constructor)

Added template function

### CODE REVIEW

File: course5/homework6-try2/homework06.cc [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+	T empty("");
+	T empty_too; 
+	std::cout << "created empty string" << std::endl;
+	assert(empty.length() == 0); 
+	assert(empty == empty_too); 
+	assert(empty == ""); 
+
+	// 2 empties should also be able to be non-compared
+	if (empty != empty_too)
+		assert(false);
```
Comment (maciekgajewski) [inline course5/homework6-try2/homework06.cc:26]: or just: ```assert(!(empty != empty_too));```

File: course5/homework6-try2/homework06.cc [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	assert(str ==  "Hello world" && str != "Hello worl"); 
+
+	std::cout << "Created empty string str='"<< empty << "' and full string str='" << str  << "'" << std::endl;
+
+	T str2 = str; // copy constructor
+	std::cout << "Created str2 from str= "<<  str2 << std::endl; 
+	assert(str == str2);
+
+	T str3;
+	str3 = "Haleluja"; // copy assignment with char*
```
Comment (maciekgajewski) [inline course5/homework6-try2/homework06.cc:39]: technically this is not "copy assignment", but rather "user-defined assignment"

File: course5/homework6-try2/homework06.cc [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+}
+
+template <typename T> 
+void test_operators()
+{
+	char test_str[] = "The Quick Brown Fox Jumps Over The Lazy Dog"; 
+	std::cout << "====== Testing operators ======" << std::endl;
+	T str(test_str);
+	T str2 = std::move(str);
+	T str3 = "Test string";
```
Comment (maciekgajewski) [inline course5/homework6-try2/homework06.cc:73]: here you can test if ```str``` is empty after the content has been stolen by ```str2```. The test may not work with std::string, thou, as one should make any assumption about the state of moved-from object.

Comment (maciekgajewski) [reply course5/homework6-try2/homework06.cc:73]: another way of testing move is to see if c_str() of str2 after the move is the same as c_str of str before the move.

File: course5/homework6-try2/mystr.cc [hunk 4]
```diff
@@ -0,0 +1,130 @@
+#include "mystr.h"
+#include <memory>
+#include <cmath> 
+#include <cstring>
+
+using namespace mystring;
```
Comment (maciekgajewski) [inline course5/homework6-try2/mystr.cc:6]: It is better to simply wrap the content of the file in a namespace. Otherwise you can get weird linking issues with free functions.

File: course5/homework6-try2/mystr.cc [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+bool MyString::operator != (const MyString& str)
+{
+	return !(*this == str);
+}
+
+bool MyString::operator != (const char* str)
+{
+	return strcmp(mStr.get(), str) != 0;
```
Comment (maciekgajewski) [inline course5/homework6-try2/mystr.cc:115]: ```std::strcmp```

File: course5/homework6-try2/homework06.cc [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+void test_basics()
+{
+	std::cout << "====== Testing basics ======" << std::endl;
+
+	T empty("");
+	T empty_too; 
+	std::cout << "created empty string" << std::endl;
+	assert(empty.length() == 0); 
+	assert(empty == empty_too); 
+	assert(empty == "");
```
Comment (maciekgajewski) [inline course5/homework6-try2/homework06.cc:22]: Add the following test please:
```
assert(std::strcmp(empty.c_str(), "") == 0);
```

File: course5/homework6-try2/homework06.cc [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+	T str6 = "World!"; 
+	std::cout << str5 << " + "<< str6 << " = " ; 
+	str5 += str6;
+	std::cout << str5 << std::endl; 
+	assert (str5 == "Hello World!");
+
+	std::cout << "Testing COntatenation. " ; 
+	T str7 = "Hello";
+	T str8 = " World Again !";
+	std::cout << str7 + str8 << std::endl;
```
Comment (maciekgajewski) [inline course5/homework6-try2/homework06.cc:123]: automate the test

File: course5/homework6-try2/homework06.cc [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	std::cout << "~~~~~ Finished Testing operators" << std::endl;
```
Comment (maciekgajewski) [inline course5/homework6-try2/homework06.cc:124]: Add automated test for ```operator<<```. Use ```std::ostringstream```. Test also for empty string.

**Review by maciekgajewski — CHANGES_REQUESTED** [review]
Mostly good!
Some tests still have to be automated. I can spot few possible bugs with empty string; tests are not covering all corner cases.
